providers/common/kdfs/hkdf.c

```diff
@@ -44,200 +55,229 @@ struct evp_kdf_impl_st {
-    kdf_hkdf_free,
-    kdf_hkdf_reset,
-    kdf_hkdf_ctrl,
-    kdf_hkdf_ctrl_str,
-    kdf_hkdf_size,
-    kdf_hkdf_derive
+static int kdf_hkdf_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    const OSSL_PARAM *p;
+    KDF_HKDF *ctx = vctx;
+    EVP_MD *md;
+    int n, got_mode = 0;
+    const char *properties = NULL;
+
+    /* Grab search properties, this should be before the digest lookup */
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_PROPERTIES))
+        != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        properties = p->data;
+    }
+    /* Handle aliasing of digest parameter names */
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_DIGEST)) != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        md = EVP_MD_fetch(PROV_LIBRARY_CONTEXT_OF(ctx->provctx), p->data,
+                          properties);
+        if (md == NULL) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
+            return 0;
+        }
+        ctx->md = md;
+    }
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_MODE)) != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        if (strcasecmp(p->data, "EXTRACT_AND_EXPAND") == 0) {
+            ctx->mode = EVP_KDF_HKDF_MODE_EXTRACT_AND_EXPAND;
+        } else if (strcasecmp(p->data, "EXTRACT_ONLY") == 0) {
+            ctx->mode = EVP_KDF_HKDF_MODE_EXTRACT_ONLY;
+        } else if (strcasecmp(p->data, "EXPAND_ONLY") == 0) {
+            ctx->mode = EVP_KDF_HKDF_MODE_EXPAND_ONLY;
+        } else {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_MODE);
+            return 0;
+        }
+        got_mode = 1;
+    }
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_MODE_INT))
+        != NULL) {
+        /* Numeric modes to be deprecated eventually */
+        if (got_mode) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_BOTH_MODE_AND_MODE_INT);
+            return 0;
+        }
+        if (!OSSL_PARAM_get_int(p, &n))
+            return 0;
+        if (n != EVP_KDF_HKDF_MODE_EXTRACT_AND_EXPAND
+            && n != EVP_KDF_HKDF_MODE_EXTRACT_ONLY
+            && n != EVP_KDF_HKDF_MODE_EXPAND_ONLY) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_MODE_INT);
+            return 0;
+        }
+        ctx->mode = n;
+    }
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_KEY)) != NULL) {
+        OPENSSL_clear_free(ctx->key, ctx->key_len);
+        ctx->key = NULL;
+        if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->key, 0,
+                                         &ctx->key_len))
+            return 0;
+    }
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_SALT)) != NULL) {
+        if (p->data_size != 0 && p->data != NULL) {
+            OPENSSL_free(ctx->salt);
+            ctx->salt = NULL;
+            if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->salt, 0,
```

Comment:
This make me sad :) Not much you can do about it here.

Reply:
The cast or the simplified code???

Reply:
the cast